zcash_primitives/src/transaction/builder.rs

```diff
@@ -107,8 +120,179 @@ impl Progress {
     }
 }
 
-enum ChangeAddress {
-    SaplingChangeAddress(OutgoingViewingKey, PaymentAddress),
+/// The balance of inputs, outputs, and fees of a transaction under construction.
+pub struct TransactionBalances {
+    total_in: Amount,
+    change_required: Amount,
+    fee_required: Amount,
+}
+
+impl TransactionBalances {
+    /// Constructs a new balance from its constituent parts. Returns an error
+    /// on overflow or underflow of amount arithmetic.
+    pub fn new(
+        total_in: Amount,
+        change_required: Amount,
+        fee_required: Amount,
+    ) -> Result<Self, BalanceError> {
+        // check for overflow and underflow at construction.
+        (total_in - (change_required + fee_required).ok_or(BalanceError::Overflow)?)
+            .ok_or(BalanceError::Underflow)?;
```

Comment:
My thinking here is that it may be necessary to compute balances for partially-constructed transactions; for example, one might supply outputs before supplying inputs, in which case the output balance could be negative. I don't think we want to constrain transaction balance at this point, since where we actually need to do that is in the builder itself.

Reply:
I narrowed the scope of TransactionBalance to just represent the balance.